### PR TITLE
[WIP] Support aarch64-pc-windows-gnu target

### PR DESCRIPTION
**Pull request by dennisameling**
This is all a bit new to me so it took me quite some hours to get to the following:

- I did `rustup default nightly-aarch64-pc-windows-msvc` - benefit of using the existing ARM64 support is that some parts in the compilation are sped up significantly as they run natively. I needed the nightly compiler because I had to do `rustc -Z unstable-options --target=x86_64-pc-windows-gnu --print target-spec-json` - not sure if I need it for further compilation steps as well or whether I can use beta/stable for that
- I opened the MSYS2 shell with `msys2_shell.cmd -use-full-path` and then did `export PATH=/clangarm64/bin:/clang64/bin:${PATH}` to ensure I have all the `clangarm64` binaries available and `clang64` as a fallback (not strictly needed I think). This ensures that MSVC can also be found as I'm cross-compiling from there for now
- In `config.toml`, I prevented the CI LLVM from being downloaded because it misses `llvm-tblgen.exe`:

```
-- Build files have been written to: C:/__/build/aarch64-pc-windows-gnu/llvm/build
running: "cmake" "--build" "." "--target" "install" "--config" "Release" "--" "-j" "8"
ninja: error: 'C:/__/build/aarch64-pc-windows-msvc/llvm/bin/llvm-tblgen.exe', needed by 'include/llvm/IR/Attributes.inc', missing and no known rule to make it
thread 'main' panicked at '
command did not execute successfully, got: exit code: 1

build script failed, must exit now', C:\Users\denni\.cargo\registry\src\github.com-1ecc6299db9ec823\cmake-0.1.44\src\lib.rs:885:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```

`config.toml`:
```
[llvm]
download-ci-llvm = false
```

- I ran into https://github.com/rust-lang/rust/issues/92163 which I could fix by removing everything in the `build` folder (except for `build\aarch64-pc-windows-msvc\llvm` because it takes ~2 hrs to build using the MSVC compiler, which can only cross-compile from x86 to arm64 and not build natively yet)
- I got to the point where it started building LLVM for `aarch64-pc-windows-gnu`, but that failed with (full log: [llvm-windows-aarch64-gnu.txt](https://github.com/dennisameling/rust/files/7792113/llvm-windows-aarch64-gnu.txt))

```
[2472/2844] Building CXX object tools/dsymutil/CMakeFiles/dsymutil.dir/SymbolMap.cpp.obj
[2473/2844] Linking CXX executable bin\bugpoint.exe
FAILED: bin/bugpoint.exe
cmd.exe /C "cd . && C:\msys64\clangarm64\bin\g++.exe -ffunction-sections -fdata-sections -Werror=date-time -Werror=unguarded-availability-new -Wall -Wextra -Wno-unused-parameter -Wwrite-strings -Wcast-qual -Wmissing-field-initializers -pedantic -Wno-long-long -Wc++98-compat-extra-semi -Wimplicit-fallthrough -Wcovered-switch-default -Wno-noexcept-type -Wdelete-non-virtual-dtor -Wsuggest-override -Wstring-conversion -Wmisleading-indentation -ffunction-sections -fdata-sections -O3 -DNDEBUG -Wl,--stack,16777216 @CMakeFiles\bugpoint.rsp -o bin\bugpoint.exe -Wl,--out-implib,lib\libbugpoint.dll.a -Wl,--major-image-version,0,--minor-image-version,0  && cd ."
ld.lld: error: duplicate symbol: vtable for llvm::FormalArgHandler
>>> defined at libLLVMPowerPCCodeGen.a(PPCCallLowering.cpp.obj)
>>> defined at libLLVMM68kCodeGen.a(M68kCallLowering.cpp.obj)
g++: error: linker command failed with exit code 1 (use -v to see invocation)
[2474/2844] Building CXX object tools/dsymutil/CMakeFiles/dsymutil.dir/MachODebugMapParser.cpp.obj
[2475/2844] Building CXX object tools/dsymutil/CMakeFiles/dsymutil.dir/DebugMap.cpp.obj
[2476/2844] Building CXX object tools/dsymutil/CMakeFiles/dsymutil.dir/dsymutil.cpp.obj
[2477/2844] Building CXX object tools/dsymutil/CMakeFiles/dsymutil.dir/MachOUtils.cpp.obj
[2478/2844] Building CXX object tools/dsymutil/CMakeFiles/dsymutil.dir/DwarfLinkerForBinary.cpp.obj
[2479/2844] Building CXX object lib/Passes/CMakeFiles/LLVMPasses.dir/PassBuilder.cpp.obj
[2480/2844] Building CXX object tools/lli/CMakeFiles/lli.dir/lli.cpp.obj
ninja: build stopped: subcommand failed.
thread 'main' panicked at '
command did not execute successfully, got: exit code: 1

build script failed, must exit now', C:\Users\denni\.cargo\registry\src\github.com-1ecc6299db9ec823\cmake-0.1.44\src\lib.rs:885:5
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
        finished in 2219.042 seconds
```

I had this in `config.toml`:

```
[target.aarch64-pc-windows-gnu]
llvm-config = "C:/msys64/clangarm64/bin/llvm-config.exe"
```

### CODE REVIEW

File: compiler/rustc_target/src/spec/windows_gnu_base.rs [hunk 1]
```diff
@@ -16,34 +16,13 @@ pub fn opts() -> TargetOptions {
         ],
     );
 
-    let mut late_link_args = LinkArgs::new();
```
Comment (dennisameling) [inline compiler/rustc_target/src/spec/windows_gnu_base.rs]: I got this from https://github.com/msys2/MINGW-packages/blob/master/mingw-w64-rust/0007-clang-subsystem.patch, but am considering to create a separate `windows_gnu_base_clang.rs` to not interfere with the existing Windows GNU-based targets. Or do we first want to switch those targets over to Clang and then add `aarch64` when that's done? I came across some code that mentions:

> Note that we only do this on MSVC as I don't think
    clang has an output mode compatible with MinGW that we need. If it does we
    should switch to clang for MinGW as well!

https://github.com/rust-lang/rust/blob/0a4dc8bc161c68320a60a6bde525adae258b6252/src/ci/scripts/install-clang.sh#L31-L35

File: src/bootstrap/bootstrap.py [hunk 2]
```diff
@@ -434,6 +434,12 @@ def download_toolchain(self, stage0=True, rustc_channel=None):
             with output(self.rustc_stamp(stage0)) as rust_stamp:
                 rust_stamp.write(key)
 
+            gcc_libs_hack = '{}/missing-libs-hack'.format(self.rust_root)
```
Comment (dennisameling) [inline src/bootstrap/bootstrap.py]: Got this from https://github.com/msys2/MINGW-packages/blob/ad72449e4dff4fb350516c1e71737ac256543499/mingw-w64-rust/PKGBUILD#L70L78. Would switching to Clang completely eliminate the need for this hack?

**Comment by dennisameling** [comment]
@mati865 do you have an idea how to work around the duplicate symbol issue mentioned in the PR description? I came across https://github.com/gabime/spdlog/issues/1800 but compiling from Ubuntu might not be something we'll want to do I guess.

File: compiler/rustc_target/src/spec/windows_gnu_base.rs (hunk 1 above)
Comment (mati865) [reply compiler/rustc_target/src/spec/windows_gnu_base.rs]: I think we we'd need special triples for x86_64 and AArch64, `windows-gnu` aren't fully compatible with Clang based toolchains. That's why we have `clang*` subsystems at MSYS2.

File: src/bootstrap/bootstrap.py (hunk 2 above)
Comment (mati865) [reply src/bootstrap/bootstrap.py]: We would have to bootstrap from Clang based toolchain instead of GCC based one.

**Comment by mati865** [comment]
> @mati865 do you have an idea how to work around the duplicate symbol issue mentioned in the PR description? I came across [gabime/spdlog#1800](https://github.com/gabime/spdlog/issues/1800) but compiling from Ubuntu might not be something we'll want to do I guess.

I have never built AArch64 LLVM and have no idea about this error (don't recall it from building on x86_64) but I would not be surprised if this is some kind of upstream LLVM bug.

**Pull request closed without merging**